pkg/importer/dataStream_test.go

```diff
@@ -7,416 +7,409 @@ import (
 
+var imageDir, _ = filepath.Abs(TestImagesDir)
+var cirrosData, _ = readFile(filepath.Join(imageDir, "cirros-qcow2.img"))
+var tinyCoreData, _ = readFile(filepath.Join(imageDir, "tinyCore.iso"))
+var testfiles = createTestData()
+var stringRdr = strings.NewReader("test data for reader 1")
+var rdrfile, _ = os.Open(filepath.Join(imageDir, "cirros-qcow2.img"))
+var testFileRdrs = bufio.NewReader(rdrfile)
+
 type fakeQEMUOperations struct {
 	e1 error
 	e2 error
 	e3 error
 }
 
-func (o *fakeQEMUOperations) ConvertQcow2ToRaw(string, string) error {
-	return o.e1
-}
+var _ = Describe("Data Stream", func() {
+	var ts *httptest.Server
 
-func (o *fakeQEMUOperations) ConvertQcow2ToRawStream(*url.URL, string) error {
-	return o.e2
-}
+	BeforeEach(func() {
+		By("[BeforeEach] Creating test server")
+		ts = createTestServer(imageDir)
+	})
 
-func (o *fakeQEMUOperations) Validate(string, string) error {
-	return o.e3
-}
+	AfterEach(func() {
+		By("[AfterEach] closing test server")
+		ts.Close()
+	})
 
-func NewFakeQEMUOperations(e1, e2, e3 error) image.QEMUOperations {
-	return &fakeQEMUOperations{e1, e2, e3}
-}
+	table.DescribeTable("with import source should", func(image, accessKeyID, secretKey string, qemu bool, want []byte, wantErr bool) {
+		if image != "" {
+			image = ts.URL + "/" + image
+		}
+		By("Creating datastream")
+		ds, err := NewDataStream(image, accessKeyID, secretKey)
+		if !wantErr {
+			Expect(err).NotTo(HaveOccurred())
+			resultBuffer := make([]byte, len(want))
+			By("Reading data from buffer")
+			ds.Read(resultBuffer)
+			Expect(ds.accessKeyID).To(Equal(accessKeyID))
+			Expect(ds.secretKey).To(Equal(secretKey))
+			Expect(ds.qemu).To(Equal(qemu))
+			Expect(reflect.DeepEqual(resultBuffer, want)).To(BeTrue())
+		} else {
+			Expect(err).To(HaveOccurred())
+		}
+	},
+		table.Entry("expect NewDataStream to succeed with valid image", "cirros-qcow2.img", "", "", true, cirrosData, false),
+		table.Entry("expect NewDataStream to fail with non existing image", "badimage.iso", "", "", false, nil, true),
+		table.Entry("expect NewDataStream to fail with invalid or missing image", "", "", "", false, nil, true),
+		table.Entry("expect NewDataStream to succeed with valid iso image", "tinyCore.iso", "accessKey", "secretKey", false, tinyCoreData, false),
+	)
+
+	It("can close all readers", func() {
+		ds, err := NewDataStream(ts.URL+"/tinyCore.iso", "", "")
+		Expect(err).NotTo(HaveOccurred())
+		emptyBuffer(ds)
+		By("Closing data stream")
+		err = ds.Close()
+		Expect(err).NotTo(HaveOccurred())
+	})
 
-func NewQEMUAllErrors() image.QEMUOperations {
-	err := errors.New("qemu should not be called from this test override with replaceQEMUOperations")
-	return NewFakeQEMUOperations(err, err, err)
-}
+	table.DescribeTable("can build selector", func(image, ep string, wantErr bool) {
+		if ep == "" {
+			ep = ts.URL
+		}
+		ds, err := NewDataStream(ep+"/"+image, "", "")
+		if !wantErr {
+			Expect(err).NotTo(HaveOccurred())
+			defer ds.Close()
```

Comment:
I think you only want to close the reader when err==nil. Maybe something like this:

Reply:
Which is what it is doing, if the Expect fails then err != nil and I don't want to close the reader

Reply:
I'm also thinking of the case where wantErr == true and NewDataStream returns no err (and thus may have created the readers).
Maybe the best way to ensure reader are close would be:
```
ds, err := NewDataStream()
if ds != nil && len(ds.Readers) > 0 {
      defer ds.Close()
}

```

Reply:
We probably should put the closing logic in the AfterEach regardless of success or failure.